### PR TITLE
Fix lagging 3d view for big data in auto color mode

### LINKED ISSUE #6397: Labels layer is lagging in 3d for big data
### 🐛 Bug Report

The original bug report is here https://forum.image.sc/t/request-for-testing-napari-0-4-19-release-candidate/87972/2?u=gbokota

In general, if user loads a big image, goes in 3D, and try to rotate, then the view lags. 

For the data from `napari-brainreg` plugin, the first example is small and does not reproduce the error on my machine. 

If I zoom that structure fills the whole space, then rotation starts working normally. 

But when I zoom out (structure is smaller) the lag is bigger and bigger. 

As I check the bug does not depend on the shader implementation (I have applied a simple shader, the same as for the image, and it is still present). 

So it looks like some obsolete calculations are performed for empty voxels. 

But I cannot find which change is responsible as whole shaders is executed under `if floatNotEqual(val, 0)` clause. 

The 3308 solves many problems so it will be  really nice to solve this issue. 

### 💡 Steps to Reproduce

1. install napari-brainreg 
2. select non-first dataset
3. change do 3d mode 
4. observe lagging interface when rotate. 

### 💡 Expected Behavior

same performance or better than in 0.4.18

### 🌎 Environment

Double check with bisect points that the bug was introduced in #3308

### 💡 Additional Context

@perlman @brisvag @jni

### PR DESCRIPTION
**Pull request by Czaki**
# References and relevant issues

closes #6397

# Description

This array fixes fps performance issues in OpenGL introduced by #3308. In that PR, the texture type was changed to float32 in order to directly pass the labels values to the texture. It turns out that OpenGL performance for float32 textures is much worse than for uint8 textures.

Here we change the code to use uint8 whenever the final number of colors is less than 255 in automatic coloring mode, or uint16 if the number is less than 65535.

This is achieved by transforming original data using a modulo-like operation that avoids the background label landing on 0.

This PR introduces numba dependency, which might not be a long-term solution. We may try to move this utility to some package that already contains compiled code. We can revisit the decision if it causes issues (such as a delay in supporting newer Python versions), and perhaps push such a function to a compiled dependency such as scikit-image.

This PR also disables the caching strategy used to speed up painting until someone actually starts painting. It is a significant speedup in startup time and reduces memory usage.

### CODE REVIEW

File: napari/layers/labels/labels.py [hunk 1]
```diff
@@ -937,10 +943,77 @@ def _partial_labels_refresh(self):
         colors_sliced = self._raw_to_displayed(
             raw_displayed, data_slice=updated_slice
         )
+        self._slice.image.view[updated_slice] = colors_sliced
```
Comment (Czaki) [inline napari/layers/labels/labels.py:952]: I do not know why I need to this line. On main the `view` array is updated, but I cannot find where. But After I add this the test are passing.

File: napari/_tests/test_windowsettings.py [hunk 2]
```diff
@@ -1,8 +1,28 @@
...
+        return None
+    return ScreenMock()
+
+
+def test_singlescreen_window_settings(make_napari_viewer, monkeypatch):
     """Test whether valid screen position is returned even after disconnected secondary screen."""
+
+    monkeypatch.setattr(
+        "napari._qt.qt_main_window.QApplication.screenAt", screen_at
+    )
```
Comment (jni) [inline napari/_tests/test_windowsettings.py]: did you mean to include these changes in this PR?

File: napari/_vispy/layers/image.py [hunk 3]
```diff
@@ -41,15 +43,20 @@ def __init__(
...
-        if ndisplay == 2:
-            return self._image_node
-        return self._volume_node
+        res = self._image_node if ndisplay == 2 else self._volume_node
+        if (
+            res.texture_format != "auto"
+            and dtype is not None
+            and _VISPY_FORMAT_TO_DTYPE[res.texture_format] != dtype
+        ):
+            raise ValueError("dtype does not match texture_format")
```
Comment (jni) [inline napari/_vispy/layers/image.py:63]: I don't understand how we could get here, since we are the ones who create all the nodes; It's also not covered by tests. So I'm wondering whether we need this?

File: napari/_vispy/layers/image.py [hunk 4]
```diff
@@ -138,13 +154,14 @@ def _on_data_change(self) -> None:
...
-            ndisplay == 2 and not isinstance(node, ImageNode)
+            ndisplay == 2
+            and not isinstance(node, ImageNode)
+            or node != self.node
         ):
             self._on_display_change(data)
         else:
             node.set_data(data)
-
-        node.visible = not self.layer._slice.empty and self.layer.visible
```
Comment (jni) [inline napari/_vispy/layers/image.py:147]: Is this change intentional? I find it surprising... Is it because `_on_display_change` correctly sets the node visibility anyway?

File: napari/_vispy/layers/labels.py [hunk 5]
```diff
@@ -59,6 +64,8 @@
     if (t == $background_value) {
         return vec4(0);
     }
+    t = t * $scale;
+    // return vec4(t,t,t,1);
```
Comment (jni) [inline napari/_vispy/layers/labels.py]: Is this left over from debugging?

File: napari/_vispy/layers/labels.py [hunk 6]
```diff
@@ -141,6 +148,9 @@
 
 """
 
+SCALE_R8 = 'float cmap(float v) { return v*255; }'
+SCALE_R16 = 'float cmap(float v) { return v*65535; }'
```
Comment (jni) [inline napari/_vispy/layers/labels.py]: Is this also left over from debugging? I don't see them used anywhere.

File: napari/_vispy/layers/labels.py [hunk 7]
```diff
@@ -158,6 +169,7 @@ def __init__(
             .replace('$use_selection', str(use_selection).lower())
             .replace('$selection', str(selection))
             .replace('$background_value', str(background_value))
+            .replace('$scale', str(scale))
```
Comment (jni) [inline napari/_vispy/layers/labels.py:165]: Debugging?

File: napari/layers/labels/labels.py [hunk 8]
```diff
@@ -1012,26 +1063,29 @@ def _raw_to_displayed(
...
         if labels_to_map.size == 0:
             return self._cached_mapped_labels[data_slice]
 
-        mapped_labels = self._to_vispy_texture_dtype(labels_to_map)
+        if self.color_mode == LabelColorMode.AUTO:
+            mapped_labels = cast_labels_to_minimum_type_auto(
+                labels_to_map, self.num_colors
+            )
+        else:  # direct
+            mapped_labels = self._to_vispy_texture_dtype(labels_to_map)
```
Comment (jni) [inline napari/layers/labels/labels.py:1084]: nice!

File: napari/layers/labels/labels.py [hunk 9]
```diff
@@ -1012,26 +1063,29 @@ def _raw_to_displayed(
...
 
-        if update_mask is not None:
-            self._cached_mapped_labels[data_slice][update_mask] = mapped_labels
-        else:
-            self._cached_mapped_labels[data_slice] = mapped_labels
+        if self._cached_labels is not None:
+            if update_mask is not None:
+                self._cached_mapped_labels[data_slice][
+                    update_mask
+                ] = mapped_labels
```
Comment (jni) [inline napari/layers/labels/labels.py:1090]: Have I mentioned I hate black? 😂

File: napari/utils/colormaps/colormap.py [hunk 10]
```diff
@@ -230,3 +235,62 @@ def default_color(self):
...
+    dtype = minimum_dtype_for_labels(num_colors + 1)
+
+    return _cast_labels_to_minimum_type_auto(data, num_colors, dtype)
+
+
+@numba.njit(parallel=True)
+def _cast_labels_to_minimum_type_auto(
+    data: np.ndarray, num_colors: int, dtype
+) -> np.ndarray:
+    result_array = np.zeros_like(data, dtype=dtype)
```
Comment (jni) [inline napari/utils/colormaps/colormap.py]: Oof. This is a big step. However, I tried my best to optimise this in NumPy, avoiding unnecessary allocations, and got 10x worse speed...

```
In [14]: %timeit cm._cast_labels_to_minimum_type_auto(data, 50, np.uint8)
OMP: Info #276: omp_set_nested routine deprecated, please use omp_set_max_active_levels instead.
33 ms ± 2.19 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [16]: %%timeit
    ...: np.add(np.mod(data, np.uint8(50), out=out, casting='unsafe'),
    ...:        np.uint8(1),
    ...:        where=(data >= 50) | (data < 0),
    ...:        out=out)
    ...:
388 ms ± 830 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Even if we optimise for unsigned ints, we are still close to 10x worse:

```
In [15]: %%timeit
    ...: np.add(np.mod(data, np.uint8(50), out=out, casting='unsafe'),
    ...:        np.uint8(1),
    ...:        where=(data >= 50),
    ...:        out=out)
    ...:
280 ms ± 458 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

So, maybe we should try to bite the bullet. Alternatively, maybe we can dispatch to the above NumPy code if numba is not available, but recommend numba (and include it in [all]).

What do you think @Czaki?

Certainly, depending on numba would open up a *lot* of optimisations. But it feels like we are making a big decision under time pressure. 😅

File: napari/utils/colormaps/colormap.py [hunk 11]
```diff
@@ -230,3 +235,62 @@ def default_color(self):
...
+    -------
+    np.ndarray
+        Casted labels data.
+    """
+    dtype = minimum_dtype_for_labels(num_colors + 1)
+
+    return _cast_labels_to_minimum_type_auto(data, num_colors, dtype)
+
+
+@numba.njit(parallel=True)
```
Comment (jni) [inline napari/utils/colormaps/colormap.py:264]: Ah, I was trying to figure out why these lines weren't covered. When I had this situation in skan, I used NUMBA_DISABLE_JIT=1 when computing coverage. (Though I can't find any evidence of this now... 😅) What do you think @Czaki?

File: napari/utils/colormaps/colormap.py [hunk 12]
```diff
@@ -230,3 +235,62 @@ def default_color(self):
...
+    Returns
+    -------
+    np.dtype
+        Minimum dtype that can hold the number of colors.
+    """
+    if num_colors <= np.iinfo(np.uint8).max:
+        return np.dtype(np.uint8)
+    if num_colors <= np.iinfo(np.uint16).max:
+        return np.dtype(np.uint16)
+    return np.dtype(np.float32)
```
Comment (jni) [inline napari/utils/colormaps/colormap.py:319]: Add tests here?

File: napari/_vispy/visuals/util.py [hunk 13]
```diff
@@ -0,0 +1,11 @@
+class TextureMinix:
```
Comment (jni) [inline napari/_vispy/visuals/util.py]:
```suggestion
class TextureMixin:
```

(And obviously the usages need to change too.)

Could you add a docstring that explains what this class does?

**Review by jni — COMMENTED** [review]
@Czaki I've left some initial comments. Still getting my head around the flow, but I think I understand the gist. It's very cool and I love that you fixed the direct case too where possible. ❤️

File: napari/utils/colormaps/colormap.py (hunk 10 above)
Comment (jni) [reply napari/utils/colormaps/colormap.py]: Ah, I just saw what you think in the PR description 😅. Yes I guess you are right that for 0.4.19 it is not a huge risk. But maybe since this is going into main, we make the dependency optional (as discussed above), and when backporting to 0.4.19 we tear out the compatibility code? Or vice versa, we can add non-numba option in a future PR, your call.

**Comment by jni** [comment]
> This PR disables caching used for speedup painting until someone starts painting. It is a significant speedup and reduces memory usage.

Nice catch!

File: napari/_vispy/layers/labels.py (hunk 5 above)
Comment (Czaki) [reply napari/_vispy/layers/labels.py]: yes

File: napari/_vispy/layers/labels.py (hunk 6 above)
Comment (Czaki) [reply napari/_vispy/layers/labels.py]: No. From alternative solution that ends as dead end, because of the lack of some API in vispy Volume node.

File: napari/_vispy/layers/labels.py (hunk 7 above)
Comment (Czaki) [reply napari/_vispy/layers/labels.py:165]: Nope. 
Vispy scale uint8 and uint16 to 0-1 range. And I need to restore original values. I have asked on vispy gitter about this, but not get any response yet.

File: napari/layers/labels/labels.py (hunk 8 above)
Comment (Czaki) [reply napari/layers/labels/labels.py:1084]: For direct, we need an analogous solution.

File: napari/layers/labels/labels.py (hunk 9 above)
Comment (Czaki) [reply napari/layers/labels/labels.py:1090]: We should either do not create so complex functions or increase line length. 

And ruff formater may behave here better.

File: napari/utils/colormaps/colormap.py (hunk 10 above)
Comment (Czaki) [reply napari/utils/colormaps/colormap.py]: I think that the best option will be to donate such code to some of our compiled dependencies (vispy, scikit?) 

Here I see a double allocation:
```
where=(data >= 50) | (data < 0),
```

I would rather open an issue for this problem than add much slower/memory consummimg code.

File: napari/_tests/test_windowsettings.py (hunk 2 above)
Comment (Czaki) [reply napari/_tests/test_windowsettings.py]: This is the fix for running a test locally on my machine. I could extract it to separate PR. But I need this for local testing as `(0,0)` is outside my screens:

obraz

File: napari/_vispy/layers/image.py (hunk 3 above)
Comment (Czaki) [reply napari/_vispy/layers/image.py:63]: To raise an exception earlier when someone will next time touch this code will not need to understand vispy error. 

(It also allows to avoid linter complaining about unused parameter)

File: napari/_vispy/layers/image.py (hunk 4 above)
Comment (Czaki) [reply napari/_vispy/layers/image.py:147]: Yes. It is intentional not to trigger obsolete events.

File: napari/_vispy/visuals/util.py (hunk 13 above)
Comment (Czaki) [reply napari/_vispy/visuals/util.py]: Is docstring ok now?

File: napari/_tests/test_windowsettings.py (hunk 2 above)
Comment (jni) [reply napari/_tests/test_windowsettings.py]: LOL that's funny. Yeah I think it might not make sense to have it in main and certainly not to have it in this PR...

File: napari/_vispy/layers/image.py (hunk 3 above)
Comment (jni) [reply napari/_vispy/layers/image.py:63]: Hmmm. That doesn't feel compelling to me... But maybe we can leave it if we add a comment, e.g.

```suggestion
            # it is a bug to hit this error — it is here to catch bugs
            # early when we are creating the wrong nodes or
            # textures for our data
            raise ValueError("dtype does not match texture_format")
```

File: napari/_vispy/layers/labels.py (hunk 7 above)
Comment (jni) [reply napari/_vispy/layers/labels.py:165]: Ah, I think that's not necessarily true — I think *maybe* there is a way to pass the values in unchanged, by setting some contrast limits to [0, 1] or something... @brisvag does this ring a bell?

File: napari/layers/labels/labels.py (hunk 9 above)
Comment (jni) [reply napari/layers/labels/labels.py:1090]: It's a black issue, not a line length issue. I would not object to:

```python
self._cached_mapped_labels[data_slice][update_mask] = (
    mapped_labels
)
```

(which comfortably fits in the line length) but black changes that back.

File: napari/_tests/test_windowsettings.py (hunk 2 above)
Comment (Czaki) [reply napari/_tests/test_windowsettings.py]: > I think it might not make sense to have it in main

Why? I think that we should go in the direction that allows run all test locally.

File: napari/utils/colormaps/colormap.py (hunk 11 above)
Comment (jni) [reply napari/utils/colormaps/colormap.py:264]: ?

File: napari/_tests/test_windowsettings.py (hunk 2 above)
Comment (jni) [reply napari/_tests/test_windowsettings.py]: Yeah but very few people have a geometry where (0, 0) is outside it? But, sure, maybe it can be discussed more extensively in a separate PR (and it would make the motivation for the code easier to discover in the future).

Comment (Czaki) [reply napari/_tests/test_windowsettings.py]: I have opened #6422

**Comment by Carreau** [comment]
restarting what looks like random failure.

**Comment by Czaki** [comment]
> restarting what looks like random failure.

I have opened PR to fix this random failures #6423

File: napari/utils/colormaps/colormap.py (hunk 11 above)
Comment (Czaki) [reply napari/utils/colormaps/colormap.py:264]: Disabling only for a given test leads to a crash. I do not want to disable it globally as it will have a big performance impact on the whole test suite...

File: napari/utils/colormaps/colormap.py (hunk 12 above)
Comment (Czaki) [reply napari/utils/colormaps/colormap.py:319]: added

File: napari/utils/colormaps/colormap.py (hunk 11 above)
Comment (Czaki) [reply napari/utils/colormaps/colormap.py:264]: ok. I go into numba code and see how they test disabling jit in test and reproduce it in test

File: napari/_vispy/layers/image.py [hunk 14]
```diff
@@ -41,15 +43,23 @@ def __init__(
             texture_format=texture_format,
         )
 
-    def get_node(self, ndisplay: int) -> Node:
+    def get_node(self, ndisplay: int, dtype=None) -> Node:
```
Comment (Carreau) [inline napari/_vispy/layers/image.py]:
```suggestion
    def get_node(self, ndisplay: int, dtype:Optional[np.dtype]=None) -> Node:
```

?

File: napari/_vispy/visuals/labels.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+from napari._vispy.visuals.util import TextureMixin
+
+SCALE_R8 = 'float cmap(float v) { return v*255; }'
+SCALE_R16 = 'float cmap(float v) { return v*65535; }'
+
+
+class LabelVisual(TextureMixin, ImageVisual):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
```
Comment (Carreau) [inline napari/_vispy/visuals/labels.py]:
```suggestion
```

Do you have to have this impl that just forward arguments, can we just remove it ? Or is there an issue with multiple inheritance?

Comment (Czaki) [reply napari/_vispy/visuals/labels.py]: Nope. This is an artifact from coding before I implemented TextureMixin. I will fix this evening.

File: napari/_vispy/layers/labels.py [hunk 16]
```diff
@@ -59,6 +64,7 @@
     if (t == $background_value) {
         return vec4(0);
     }
+    t = t * $scale;
```
Comment (jni) [inline napari/_vispy/layers/labels.py:66]:
```suggestion
    // VisPy automatically scales uint8 and uint16 to [0, 1].
    // this line fixes returns values to their original range.
    t = t * $scale;
```

File: napari/_vispy/visuals/util.py [hunk 17]
```diff
@@ -0,0 +1,17 @@
+class TextureMixin:
+    """
+    This mixin is created to avoid using vispy's private api.
+    vispy stores texture format `node._texture.internalformat`
+    And we should not relay on it if it is not needed.
+    """
```
Comment (jni) [inline napari/_vispy/visuals/util.py]:
```suggestion
    """Store texture format passed to VisPy classes.
    
    We need to refer back to the texture format, but VisPy
    stores it in a private attribute — ``node._texture.internalformat``.
    This mixin is added to our Node subclasses to avoid having to
    access private VisPy attributes.
    """
```

File: napari/_vispy/visuals/util.py [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+
+    def __init__(self, *args, texture_format, **kwargs):
+        super().__init__(*args, texture_format=texture_format, **kwargs)
+        # save texture format to not depend on vispy's private api
+        self.unfreeze()
+        self._texture_format = texture_format
+        self.freeze()
```
Comment (jni) [inline napari/_vispy/visuals/util.py]:
```suggestion
        # classes using this mixin may be frozen dataclasses.
        # we save the texture format between unfreeze/freeze.
        self.unfreeze()
        self._texture_format = texture_format
        self.freeze()
```

File: napari/layers/labels/labels.py (hunk 1 above)
Comment (jni) [reply napari/layers/labels/labels.py:952]: @Czaki which tests? This of course reminds me of #6112, and indeed [this comment](https://github.com/napari/napari/pull/6112/files#r1291613760) by @andy-sweet:

> If needed, I think you could fix this fairly easily by doing something like `self._slice.image.view[displayed_indices] = self._raw_to_displayed(self._slice.image.raw[displayed_indices])` (or maybe just `:` instead of `displayed_indices` depending on how `_raw_to_displayed` works). Either here or in `_partial_labels_refresh`.

So I guess it *was* needed, 😅 but it would be good to keep a note about it here, given #6185. So:

```suggestion
        # This line is needed for test_... to pass
        # See https://github.com/napari/napari/pull/6112/files#r1291613760
        # and https://github.com/napari/napari/pull/6185
        self._slice.image.view[updated_slice] = colors_sliced
```

File: napari/layers/labels/labels.py [hunk 19]
```diff
@@ -937,10 +943,77 @@ def _partial_labels_refresh(self):
...
+        # Remove the latest one-pixel border from the result
+        delta_slice = tuple(
+            slice(s1.start - s2.start, s1.stop - s2.start)
+            for s1, s2 in zip(data_slice, expanded_slice)
+        )
+        return sliced_labels[delta_slice]
+
+    def _get_cache_dtype(self) -> np.dtype:
+        if self.color_mode == LabelColorMode.DIRECT:
+            return np.dtype(np.float32)
```
Comment (jni) [inline napari/layers/labels/labels.py:1003]: I'm confused — don't we switch texture types based on the values in the color dictionary?

File: napari/layers/labels/labels.py (hunk 8 above)
Comment (jni) [reply napari/layers/labels/labels.py:1084]: Wait — this clause *is* for direct?

File: napari/utils/colormaps/colormap.py [hunk 20]
```diff
@@ -230,3 +235,62 @@ def default_color(self):
...
+    """
+    dtype = minimum_dtype_for_labels(num_colors + 1)
+
+    return _cast_labels_to_minimum_type_auto(data, num_colors, dtype)
+
+
+@numba.njit(parallel=True)
+def _cast_labels_to_minimum_type_auto(
+    data: np.ndarray, num_colors: int, dtype
+) -> np.ndarray:
```
Comment (jni) [inline napari/utils/colormaps/colormap.py]: I think we might as well make this function have a generic style here already?

```suggestion
def modulo_plus_one(
    values: np.ndarray, n: int, dtype
) -> np.ndarray:
    """Like ``array % n`` but with 1 added to result for values >n.

    This ensures (1) an output value in [0, n] (inclusive), and (2) that
    no nonzero values in the input are zero in the output.
    """
```

File: napari/utils/colormaps/colormap.py [hunk 21]
```diff
@@ -230,3 +235,62 @@ def default_color(self):
...
+    result_array = np.zeros_like(data, dtype=dtype)
+
+    # iterate over data and calculate modulo num_colors assigning to result_array
+
+    for i in numba.prange(data.size):
+        if num_colors > data.flat[i] >= 0:
```
Comment (jni) [inline napari/utils/colormaps/colormap.py]:
```suggestion
        if 0 <= data.flat[i] < num_colors:
```

This is slightly more readable? 😅

File: napari/utils/colormaps/_tests/test_colormap.py [hunk 22]
```diff
@@ -110,3 +113,33 @@ def test_mapped_shape(ndim):
...
+def test_minimum_dtype_for_labels(num, dtype):
+    assert colormap.minimum_dtype_for_labels(num) == dtype
+
+
+@pytest.fixture()
+def disable_jit(monkeypatch):
+    with patch("numba.core.config.DISABLE_JIT", True):
+        importlib.reload(colormap)
+        yield
+    importlib.reload(colormap)  # rever to original state
```
Comment (jni) [inline napari/utils/colormaps/_tests/test_colormap.py]: Lovely!

```suggestion
    importlib.reload(colormap)  # revert to original state
```

**Review by jni — CHANGES_REQUESTED** [review]
@Czaki I've left minor suggestions. I'm still unsatisfied with the flow in `_raw_to_displayed` but that is not your mess. 😂 I don't think we need to fix it in this PR.

I'm also unsure about the state of the direct map — different lines suggest different things with regards to the texture format. Are direct colours always 32-bit or not? It seems we are very close to fixing, so it would be good to do in this PR also...

Thanks for tackling this!

**Comment by Czaki** [comment]
> I'm also unsure about the state of the direct map — different lines suggest different things with regards to the texture format. Are direct colours always 32-bit or not? It seems we are very close to fixing, so it would be good to do in this PR also...

For direct colormap, we need to implement separate functions for transforming labels to compressed format based on color dict, not num of colors. I could start working on this. But it will be a separate track.

File: napari/_vispy/layers/labels.py [hunk 23]
```diff
@@ -473,11 +481,17 @@ def _on_colormap_change(self, event=None):
         mode = self.layer.color_mode
 
         if mode == 'auto':
+            dtype = minimum_dtype_for_labels(self.layer.num_colors)
+            if issubclass(dtype.type, np.integer):
+                scale = np.iinfo(dtype).max
+            else:
```
Comment (Czaki) [inline napari/_vispy/layers/labels.py]:
```suggestion
            else:  # float32 texture
```

File: napari/layers/labels/labels.py (hunk 19 above)
Comment (Czaki) [reply napari/layers/labels/labels.py:1003]: No. Before this PR everything goes to float32. This PR fixes only Auto color mode

File: napari/layers/labels/labels.py (hunk 8 above)
Comment (Czaki) [reply napari/layers/labels/labels.py:1084]: Yes. This is old behavior for all textures, now kept for direct mode.

File: napari/utils/colormaps/colormap.py (hunk 20 above)
Comment (Czaki) [reply napari/utils/colormaps/colormap.py]: This is intentionally private function for hiding implementation details. 

Public API is in `cast_labels_to_minimum_type_auto`. Using this directly may be not a proper idea, as I do not add any guard about dtype.

Comment (jni) [reply napari/utils/colormaps/colormap.py]: It can be private, but I think it's useful to rename it as above and add a proper docstring, as it makes it easier to understand the logic.

File: napari/layers/labels/labels.py (hunk 1 above)
Comment (jni) [reply napari/layers/labels/labels.py:952]: @Czaki 👆?

**Comment by jni** [comment]
@Czaki thanks for the clarifications! Just two more comments and then I think it's ready to go:

- [this comment about updating the slice view](https://github.com/napari/napari/pull/6411/files#r1390182283): would be good to add references in the code since we're not really sure why it's needed and also probably needs to be fixed with better architecture in the near future.
- `modulo_plus_one` can of course be private for now (`_modulo_plus_one`), but [I think it should still be renamed](https://github.com/napari/napari/pull/6411#discussion_r1390312673) and include a proper docstring to understand what it's doing and why.

Then I think we can merge this. Will look at #6439 now. 🙏

**Review by jni — APPROVED** [review]

File: napari/utils/colormaps/colormap.py [hunk 24]
```diff
@@ -230,3 +235,78 @@ def default_color(self):
...
+    Returns
+    -------
+    np.ndarray
+        The result: ``values`` for values in [0, n), ``values % n + 1``
+        everywhere else.
+    """
+    result = np.empty_like(values, dtype=dtype)
+
+    for i in numba.prange(values.size):
+        if 0 <= values.flat[i] < n:
```
Comment (Czaki) [inline napari/utils/colormaps/colormap.py]:
```suggestion
        if 0 <= values.flat[i] <= n:
```

for consistency with docstring

Comment (jni) [reply napari/utils/colormaps/colormap.py]: @Czaki Ah, well the docstring is inconsistent with itself: the one-liner says `for values >n`, but the result description says `values in [0, n)` are not changed. So we can:
- update the oneliner to say `>=n`.
- Make this change and also update the result description.

What do we prefer? In the first scenario (and current code), `n-1` gets mapped to itself, `n` gets mapped to 1, `2*n - 1` gets mapped to `n`, and `2*n` gets mapped to 1. In the alternate scenario, `n` gets mapped to itself, and `n+1` gets mapped to 2. Either way, there is a discontinuity at one side of `n`.

Since `<` is slightly faster to evaluate than `<=` (? maybe this is no longer true with modern compilers), I suggest I just change the one-liner.

Comment (jni) [reply napari/utils/colormaps/colormap.py]: I pushed up the updated description. But let me know if you want me to change it in the other direction.

**Comment by Czaki** [comment]
I spot one problem. There is a lag when someone starts drawing (because of creating a cache array). Should we make it on the selection of paint mode?

**Comment by jni** [comment]
@Czaki I was about to push some changes to the `_modulo_plus_one` docstring (which needs updating now of course). And I was a bit unhappy with low label values now being changed in the node data, but understood the need.

However, now that I am looking at it, it doesn't look like we actually even support a non-zero background label through our public API? (There is no relevant kwarg to the Labels layer.) (I just found that you could set layer.colormap.background_value, but a GitHub code search finds **no** instances of anyone touching that value.) So, I think rather than support it through this whole stack *and* codify it in our tests (`_data_view` is an implementation detail that will *have* to disappear once we move slicing information off the layers, which is absolutely necessary for multicanvas), I think we should just revert your last few commits and not worry about nonzero background labels until someone complains. What do you think?

Anyway, I'll push my changes, they can be reverted just as well; but I really think we should strongly consider reverting them because even temporary test values add significant code debt — in the future someone will have to think about whether those tests need to be ported to some new system, or just discarded, or something else...

**Comment by Czaki** [comment]
> However, now that I am looking at it, it doesn't look like we actually even support a non-zero background label through our public API? (There is no relevant kwarg to the Labels layer.) (I just found that you could set layer.colormap.background_value,

My impression from the previous talk (during the last fixing of labels rendering) is that we want to support setting different backgrounds from 0.

**Review by jni — APPROVED** [review]

**Comment by brisvag** [comment]
> My impression from the previous talk (during the last fixing of labels rendering) is that we want to support setting different backgrounds from 0. Maybe we should just drop it. I just do not want to break this.

I remember this being added specifically cause someone asked for it and they had data with -1 as background... I think @kevinyamauchi was implementing this at the time maybe?

**Comment by jni** [comment]
Good thing @Czaki convinced me to keep the functionality then. 😅 Thanks for chiming in @brisvag! 🙏

**Comment by jni** [comment]
Gaaah, @Czaki while I was rewording the description, I thought of another issue with this PR: we should not use num_colors to do the mod operation, or labels that have the same color will *always* have the same color. The mod operation (a) should only occur when the label data dtype doesn't fit in uint8 in the first place, and (b) should be by a prime number that is greater than num_colors but still fits in the desired dtype, *and* that prime number should get changed when Shuffle is hit. 😕 And we should figure out a test that catches this issue in the future...

At any rate, though, in order to avoid complicating this PR further, I think that we should merge this, and fix it in a subsequent PR. We can raise an issue with the 0.4.19 milestone to make sure it gets handled...

Have I got the analysis right?

**Comment by jni** [comment]
Ok after a quick chat with @Czaki we have determined that this issue also exists on main. Therefore, I'm going to merge this and make an issue, which we can then fix. (It may even go back further, but I'm not certain about this, will need to test properly.)

**Comment by jni** [comment]
The issue is explained in detail in #6448.
